### PR TITLE
Carbon aware CLI update

### PR DESCRIPTION
**Pull request by pritipath**
Issue Number: 172

## Summary
Refactored the CLI code to consume the new Plugin Interface

## Changes

Modified Progam.cs to use Dependency Injection in order to access ICarbonAware service
Updated CarbonAwareCLI.cs to call the new interface and API

## Checklist

- [x] Local Tests Passing?
- [x] CICD and Pipeline Tests Passing?
- [x] Added any new Tests?
- [x] Documentation Updates Made?
- [x] This is not a breaking change. If it is, please describe it below.

## Is this a breaking change? No
If yes, what workflow does this break? 

## Anything else?
Other comments, collaborators, etc.
This PR Closes Issue #

### CODE REVIEW

File: src/dotnet/CarbonAware.CLI.Tests/CarbonAwareCLITests.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        
+        var cli = new CarbonAwareCLI(args, It.IsAny<ICarbonAwareAggregator>());
+
+        Assert.AreEqual("test", cli._state.Locations[0]); 
+        Assert.AreEqual(DateTime.Parse("2021-11-11"), cli._state.Time); 
+        Assert.AreEqual(DateTime.Parse("2021-12-12"), cli._state.ToTime); 
+    }
+
+    [Test]
+    public void ParseCommandLineArguments_ThorwsErrorWhenLocationNotProvided()
```
Comment (juzuluag) [inline src/dotnet/CarbonAware.CLI.Tests/CarbonAwareCLITests.cs:47]: nit: typo `...Throws...`

File: src/dotnet/CarbonAware.CLI/CarbonAware.CLI.csproj [hunk 2]
```diff
@@ -33,11 +33,12 @@
 
   <ItemGroup>
     <ProjectReference Include="..\CarbonAware.Plugins.BasicJsonPlugin\CarbonAware.Plugins.BasicJsonPlugin.csproj" />
+    <ProjectReference Include="..\CarbonAware.Plugins.JsonReaderPlugin\CarbonAware.Plugins.JsonReaderPlugin.csproj" />
```
Comment (juzuluag) [inline src/dotnet/CarbonAware.CLI/CarbonAware.CLI.csproj:36]: This can be removed.

File: src/dotnet/CarbonAware.CLI/Program.cs [hunk 3]
```diff
@@ -1,14 +1,42 @@
...
+using CarbonAware.Aggregators.Configuration;
+using Microsoft.Extensions.DependencyInjection;
 
 class Program
 {
-    static void Main(string[] args)
-    {
-        var cli = new CarbonAwareCLI(args);
+    private static ICarbonAwareAggregator _aggregator;
+    public static void Main(string[] args)
```
Comment (juzuluag) [inline src/dotnet/CarbonAware.CLI/Program.cs:10]: you could defined `Main` as `static async Task<int> Main(string[] args)`, so you don't need to do `emissions.Result`. As a general comment, mixing sync with async is not recommended.

File: src/dotnet/CarbonAware.CLI/Program.cs [hunk 4]
```diff
@@ -1,14 +1,42 @@
...
+
         if (cli.Parsed)
         {
             var emissions = cli.GetEmissions();
-            cli.OutputEmissionsData(emissions);
+            cli.OutputEmissionsData(emissions.Result);
         }
     }
+
+    private static void InitializePlugin() {
```
Comment (juzuluag) [inline src/dotnet/CarbonAware.CLI/Program.cs:23]: Maybe this should return the `Aggregator`. Also renaming the function to accommodate what it does would be very helpful.

File: src/dotnet/CarbonAware.CLI/Program.cs [hunk 5]
```diff
@@ -1,14 +1,42 @@
...
+        serviceCollection.AddCarbonAwareEmissionServices();
+        serviceCollection.AddLogging();     
+        var serviceProvider = serviceCollection.BuildServiceProvider();
+        var services = serviceProvider.GetServices<ICarbonAwareAggregator>();
+        
+        //Currently there is just one implementation. This will have to change once we implement WattTime
+        _aggregator = services.First();
+    }
+    private async Task GetEmissionsData(string[] args) {
+        var cli = new CarbonAwareCLI(args, _aggregator);
```
Comment (juzuluag) [inline src/dotnet/CarbonAware.CLI/Program.cs:34]: Why 2 `new CarbonAwareCLI` calls?

File: src/dotnet/CarbonAware.CLI/Program.cs [hunk 6]
```diff
@@ -1,14 +1,35 @@
...
+        //Currently there is just one implementation. This will have to change once we implement WattTime
+        return services.First();
+    }
+    private static async Task GetEmissionsData(string[] args, ICarbonAwareAggregator aggregator) {
+        var cli = new CarbonAwareCLI(args, aggregator);
+
         if (cli.Parsed)
         {
-            var emissions = cli.GetEmissions();
+            var emissions = await cli.GetEmissions();
```
Comment (rakker91) [inline src/dotnet/CarbonAware.CLI/Program.cs:31]: If we're making this asynchronous, we should consider changing the method signature to be GetEmissionsAsync

**Review by rakker91 — CHANGES_REQUESTED** [review]

**Review by rakker91 — APPROVED** [review]

**Review by juzuluag — APPROVED** [review]
LGTM. Great work

**Pull request merged**